customer.h

```diff
@@ -1,23 +1,33 @@
 #ifndef CUSTOMER_H
 #define CUSTOMER_H
 
-#include <iostream>
 #include <thread>
-#include <chrono>
-
-using namespace std;
+#include <future>
 
 class Shop;
+class Semaphore;
 
 class Customer
 {
 public:
     Customer(std::string name, Shop* shop);
+    void joinThread();
+    bool isTerminated() const { return mTerminated; }
+    bool stopRequested();
+    void stop();
 
 private:
     std::string mCustomerName;
     std::thread mCustomerThread;
+
+    std::promise<void> exitSignal;
+    std::future<void> futureObj;
+
     Shop* pBelongsToShop;
+    Semaphore* mBarberNotifier;
+    Semaphore* mCustomersNotifier;
+
+    bool mTerminated;
```

Comment:
Is this accessed by many threads? Is it safe? Maybe this would be a good candidate for std::atomic. But I am not sure whether it is actually needed.

Reply:
It is used in Shop::clearFirstTerminatedCustomerFound() function (accessed only in Shop, thus the main thread) in order to remove a Customer object that has been served (in order to avoid an increasing vector) from std::vector that holds references to created Customers.

Reply:
Declared it as atomic.